overlord/snapstate/snapmgr.go

```diff
@@ -551,6 +556,45 @@ func (m *SnapManager) ensureUbuntuCoreTransition() error {
 	return nil
 }
 
+// ensureSnapConfineApparmor ensures that we have a valid apparmor
+// profile for snap-confine when we are in re-exec mode
+func (m *SnapManager) ensureSnapConfineApparmor() error {
+	if !release.OnClassic {
+		return nil
+	}
+
+	root := filepath.Join(dirs.SnapMountDir, "/core/current/")
+	snapConfineInCore, err := filepath.EvalSymlinks(filepath.Join(root, "/usr/lib/snapd/snap-confine"))
+	if err != nil {
+		return err
+	}
+
+	apparmorProfilePath := filepath.Join(dirs.SystemApparmorDir, strings.Replace(snapConfineInCore[1:], "/", ".", -1))
+	if osutil.FileExists(apparmorProfilePath) {
+		return nil
+	}
+
+	// FIXME2: cleanup old entries here
+
+	// FIXME: what to do for debian? debian has a less powerful apparmor
+	//        so we can not use the ubuntu profile?
+	apparmorProfile, err := ioutil.ReadFile(filepath.Join(root, "/etc/apparmor.d/usr.lib.snapd.snap-confine"))
+	if err != nil {
+		return err
+	}
+	apparmorProfileForCore := strings.Replace(string(apparmorProfile), "/usr/lib/snapd/snap-confine", snapConfineInCore, -1)
+
+	if err := ioutil.WriteFile(apparmorProfilePath, []byte(apparmorProfileForCore), 0644); err != nil {
+		return err
+	}
+
+	if output, err := exec.Command("apparmor_parser", "--replace", "--write-cache", apparmorProfilePath).CombinedOutput(); err != nil {
```

Comment:
Those options are less interesting for a single profile like snap-confine where all this happens as part of the snap refresh. -O no-expr-simplify is important on ARM with a lot of profiles that use our template, etc, but the gain will not be significant for this one profile. IMHO, this can be fixed up in a future commit if it keeps this PR simple.